### ISSUE TITLE
Provide a default NullSessionIdManager implementation

### ISSUE THREAD

**Issue by markkolich** [eclipse/jetty.project#3963]
`jetty-9.4.19.v20190610`
`openjdk version "12.0.2" 2019-07-16`

I've got a high-performance stateless web-service running on `jetty-9.4.19.v20190610` in which I've explicitly disabled "sessions" by registering the Jetty provided beans with my embedded server:

- [NullSessionCache](https://github.com/eclipse/jetty.project/blob/02c247be5fd2b70ea0640f7355132a4efdbbab1f/jetty-server/src/main/java/org/eclipse/jetty/server/session/NullSessionCache.java)
- [NullSessionDataStore](https://github.com/eclipse/jetty.project/blob/02c247be5fd2b70ea0640f7355132a4efdbbab1f/jetty-server/src/main/java/org/eclipse/jetty/server/session/NullSessionDataStore.java)

On finding these, I went looking for a `NullSessionIdManager` in `jetty-server` but couldn't find one.

Is it intentional that a `NullSessionIdManager` doesn't exist?

I ended up wiring in my own `NoopSessionIdManager implements SessionIdManager` which is just a bare implementation of the `SessionIdManager` interface. But, it would be great if Jetty provided such a default implementation for me in `jetty-server`, given that it is already providing `NullSessionCache` and `NullSessionDataStore`.

Relevant snippet:

```java
final Server server = new Server();

server.addBean((SessionCacheFactory) NullSessionCache::new);
server.addBean((SessionDataStoreFactory) handler -> new NullSessionDataStore());

// Use our own no-op session ID manager to prevent Jetty from spawning its own.
// Could be provided by Jetty?
server.setSessionIdManager(new NoopSessionIdManager());
```

That said though, is this the right way to disable sessions? I'm a bit surprised I can't just set a flag on my `Server` instance to stop Jetty from spawning its own `DefaultSessionIdManager`?

**Comment by janbartel** [comment]
Hi Mark,

The NullSessionCache and the NullSessionDataStore don't actually disable sessions.  The former ensures that all requests operate on distinct copies of the session. So you can use it in front of any of the SessionDataStores.  To create sessions that are only in memory, you would use the NullSessionDataStore.


If you're not using sessions (and none of your 3rd party libs do either), then the DefaultSessionIdManager will never be called, so no harm to have it there, just remember to disable session scavenging by (setting 0 on the Housekeeper used by the DefaultSessionIdManager).  

It sounds more like you want request.getSession to throw some kind of error to indicate you don't want sessions to be used?

**Comment by gregw** [comment]
@markkolich if you don't want sessions and have an embedded server then best to just not have a SessionHandler at all!   If you are not using servlet security then you can drop the SecurityHandler as well!

How are you creating your context?

**Comment by markkolich** [comment]
Thanks for the pointers, @janbartel and @gregw.

I'm definitely not using sessions of any kind in my embedded server, nor via any 3rd party libraries. I have zero calls to `HttpServletRequest#getSession()`. My service is as bare-bones as it gets.

Are you saying I can just set a `null` `SessionHandler` and `null` `SecurityHandler` on my `WebAppContext` to disable all of this session stuff?

```java
final WebAppContext context = new WebAppContext();

context.setContextPath("/foo");

// Intentionally skip scanning JARs for Servlet 3 annotations.
context.setAttribute(WebInfConfiguration.WEBINF_JAR_PATTERN, "^$");
context.setAttribute(WebInfConfiguration.CONTAINER_JAR_PATTERN, "^$");

// Don't want JSP support, either.
context.setDefaultsDescriptor(null);

context.setSessionHandler(null); // <<--??
context.setSecurityHandler(null); // <<--??
```

🤔

**Comment by gregw** [comment]
Are you using a web.xml?   If not, then don't even use a WebAppContext, use a ServletContextHandler instead.    SessionHandler and SecurityHandler are optional, only have them if needed.... although they may be a little less optional in a WebAppContext, depending on the Configurations that you have.

**Comment by markkolich** [comment]
> Are you using a web.xml?

Yes, I am.

Apologies, I mistakenly omitted that in my earlier comment.

I'm pointing my `WebAppContext` at a WAR via `context.setWar(...)` when setting up the server.

That said, as you somewhat hinted, the `WebAppContext` was not happy with `context.setSessionHandler(null)` and `context.setSecurityHandler(null)`:

```
Exception in thread "main" java.lang.NullPointerException
  at org.eclipse.jetty.servlet.ServletContextHandler.setSessionHandler(ServletContextHandler.java:604)
```

![image](https://user-images.githubusercontent.com/1202420/62916405-2efd6280-bd4d-11e9-8a54-8baf5a4098a7.png)

In fact, this entire thread just reminded me of https://github.com/eclipse/jetty.project/issues/1809 from a few years ago in which I also tried `ServletContextHandler.NO_SESSIONS|ServletContextHandler.NO_SECURITY` which didn't work either.

**Comment by gregw** [comment]
Hmmm let me see if I can fix that NPE....

**Comment by gregw** [comment]
@markkolich I just got it to work on our test async-rest webapp with the following code:
```
        WebAppContext webapp = new WebAppContext(server, "/", null, null, null, null, 0);
        File warFile = JettyDistribution.resolve("demo-base/webapps/async-rest.war").toFile();
        webapp.setWar(warFile.getAbsolutePath());
```

**Comment by gregw** [comment]
... and the dump shows that it worked:
```
+= o.e.j.w.WebAppContext@6aba2b86{Async REST Webservice Example,/,[file:///tmp/jetty-0_0_0_0-8080-async-rest_war-_-any-5955600231072177317.dir/webapp/, jar:file:///tmp/jetty-0_0_0_0-8080-async-rest_war-_-any-5955600231072177317.dir/webapp/WEB-INF/lib/example-async-rest-jar-9.4.20-SNAPSHOT.jar!/META-INF/resources],AVAILABLE}{/usr/local/home/gregw/src/jetty-9.4.x/jetty-distribution/target/distribution/demo-base/webapps/async-rest.war} - STARTED
|  +~ org.eclipse.jetty.jmx.MBeanContainer@7193666c
|  += ServletHandler@557caf28{STARTED} - STARTED
```

**Comment by markkolich** [comment]
@gregw OK, yeah, you're totally right. This works just fine:

```java
import static org.eclipse.jetty.servlet.ServletContextHandler.*;

final WebAppContext context = new WebAppContext(server, context_, null, null,
  null, null, NO_SESSIONS|NO_SECURITY);
```

Not sure what was going on in https://github.com/eclipse/jetty.project/issues/1809 but setting `NO_SESSIONS|NO_SECURITY` on a `WebAppContext` with `9.4.19.v20190610` definitely disables the default session and security handlers.

Thanks for your help @gregw and @janbartel!
